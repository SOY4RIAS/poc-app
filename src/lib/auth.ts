import type {
  GetServerSidePropsContext,
  NextApiRequest,
  NextApiResponse,
} from 'next';
import type { NextAuthOptions } from 'next-auth';
import { getServerSession } from 'next-auth';
import CredentialsProvider from 'next-auth/providers/credentials';

export const authConfig = {
  secret: process.env.NEXTAUTH_SECRET,
  session: {
    strategy: 'jwt',
  },
  providers: [
    CredentialsProvider({
      name: 'Credentials',
      credentials: {
        email: {
          label: 'Email',
          type: 'text',
          placeholder: 'Enter email',
        },
        password: { label: 'Password', type: 'password' },
      },

      // @ts-ignore
      async authorize(credentials) {
        if (!credentials?.email || !credentials?.password) {
          return Promise.resolve(null);
        }

        return Promise.resolve({
          id: 1,
          name: credentials?.email?.split('@')[0],
          email: credentials?.email,
          password: credentials?.password,
        });
      },
    }),
  ],
} satisfies NextAuthOptions;

export function auth(
  ...args:
    | [GetServerSidePropsContext['req'], GetServerSidePropsContext['res']]
    | [NextApiRequest, NextApiResponse]
    | []
) {
  return getServerSession(...args, authConfig);
}
