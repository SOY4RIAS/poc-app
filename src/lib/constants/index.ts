export { Paths } from './paths';
export { HttpStatus } from './httpStatus';
