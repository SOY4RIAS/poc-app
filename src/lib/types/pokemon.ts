export interface GetPokemonsResponse {
  count: number;
  next: string | null;
  previous: string | null;
  results: Pokemon[];
}

export interface GetPokemonsParams {
  offset?: string;
  limit?: string;
}

export interface Pokemon {
  name: string;
  url: string;
}
