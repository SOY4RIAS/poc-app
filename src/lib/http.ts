import { HTTP_STATUS_MESSAGE } from '@/lib/constants/httpStatus';

export interface HttpResponse<T> {
  data: T;
  status: string;
  statusCode: number;
}

export class HttpClient {
  constructor(private readonly baseUrl: string) {
    if (!baseUrl) {
      throw new Error('baseUrl is required');
    }
  }

  async jsonResponse<T>(response: Promise<Response>) {
    const awaitedResponse = await response;

    return {
      data: (await awaitedResponse.json()) as T,
      statusCode: awaitedResponse.status,
      status: HTTP_STATUS_MESSAGE[awaitedResponse.status],
    };
  }

  async get<T>(
    path: string,
    params?: URLSearchParams
  ): Promise<HttpResponse<T>> {
    const url = new URL(`${this.baseUrl}${path}`);
    if (params) {
      url.search = params.toString();
    }
    const fetchPromise = fetch(url);

    return this.jsonResponse<T>(fetchPromise);
  }

  async post<T, B>(path: string, body: B): Promise<HttpResponse<T>> {
    const fetchPromise = fetch(`${this.baseUrl}${path}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });

    return await this.jsonResponse<T>(fetchPromise);
  }
}
