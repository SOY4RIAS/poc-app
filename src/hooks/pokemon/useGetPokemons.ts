import { useQuery } from '@tanstack/react-query';

import { getPokemons } from '@/services/pokemonService';
import { GetPokemonsResponse } from '@/lib/types';
import { HttpResponse } from '@/lib/http';

interface UseGetPokemonsParams {
  search?: string;
  initialData?: HttpResponse<GetPokemonsResponse>;
}

export const useGetPokemons = ({
  initialData,
  search,
}: UseGetPokemonsParams) => {
  return useQuery({
    queryKey: ['pokemons', search],
    queryFn: () => getPokemons(search),
    initialData: initialData || undefined,
  });
};
