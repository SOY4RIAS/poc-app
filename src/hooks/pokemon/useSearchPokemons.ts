import { useQuery } from '@tanstack/react-query';

import { getPokemons } from '@/services/pokemonService';
import { GetPokemonsResponse } from '@/lib/types';
import { HttpResponse } from '@/lib/http';

interface UseGetPokemonsParams {
  search?: string;
  initialData?: HttpResponse<GetPokemonsResponse>;
}

export const useSearchPokemons = ({
  initialData,
  search,
}: UseGetPokemonsParams) => {
  const query = useQuery({
    queryKey: ['pokemons'],
    queryFn: () => {
      return getPokemons(new URLSearchParams({ limit: '100000' }).toString());
    },
    initialData: initialData || undefined,
    enabled: !!search,
    select: (response) => {
      return response.data.results
        .filter((pokemon) => {
          return pokemon.name
            .toLocaleLowerCase()
            .includes(search?.trim()?.toLocaleLowerCase() || '');
        })
        .slice(0, 10);
    },
  });

  return query;
};
