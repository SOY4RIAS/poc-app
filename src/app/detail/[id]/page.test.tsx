import { render } from '@testing-library/react';

import Detail from '@/app/detail/[id]/page';
import { resolvedComponent } from '@/lib/tests';
import { pokemonMock } from '@/lib/tests/mocks/pokemon';

jest.mock('@/services/pokemonService', () => ({
  getPokemon: jest.fn(() => Promise.resolve(pokemonMock)),
}));

describe('DetailPage', () => {
  it('should render', async () => {
    const ResolvedComponent = await resolvedComponent(Detail, {
      params: {
        id: '1',
      },
    });
    const { container } = render(<ResolvedComponent />);
    expect(container).toMatchSnapshot();
  });
});
