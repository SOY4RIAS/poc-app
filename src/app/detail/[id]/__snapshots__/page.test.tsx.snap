// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailPage should render 1`] = `
<div>
  <main
    class="flex h-full flex-col items-center justify-center"
  >
    <div
      class="rounded-lg border bg-card text-card-foreground shadow-sm w-full max-w-3xl"
    >
      <div
        class="flex flex-col space-y-1.5 p-6"
      >
        <h3
          class="text-2xl font-semibold leading-none tracking-tight"
        >
          bulbasaur
        </h3>
      </div>
      <div
        class="p-6 pt-0"
      >
        <div
          class="flex items-center space-x-4"
        >
          <img
            alt="bulbasaur"
            class="border aspect-square rounded-xl"
            data-nimg="1"
            decoding="async"
            height="96"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fexample.com%2Fbulbasaur.gif&w=256&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fexample.com%2Fbulbasaur.gif&w=96&q=75 1x, /_next/image?url=https%3A%2F%2Fexample.com%2Fbulbasaur.gif&w=256&q=75 2x"
            style="color: transparent;"
            width="96"
          />
          <div
            class="grid gap-1.5"
          >
            <div
              class="flex items-center space-x-2"
            >
              <span
                class="text-lg font-semibold"
              >
                Type
              </span>
              <div
                class="flex items-center space-x-2 uppercase"
              >
                grass
              </div>
            </div>
            <div
              class="flex items-center space-x-2"
            >
              <span
                class="text-lg font-semibold"
              >
                Abilities
              </span>
              <ul
                class="flex space-x-2 text-sm"
              >
                <li>
                  overgrow
                </li>
                <li>
                  chlorophyll
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="grid gap-1.5 mt-5"
        >
          <div
            class="flex items-center space-x-2"
          >
            <span
              class="text-lg font-semibold"
            >
              Height
            </span>
            <span>
               cm
            </span>
          </div>
          <div
            class="flex items-center space-x-2"
          >
            <span
              class="text-lg font-semibold"
            >
              Weight
            </span>
            <span>
               g
            </span>
          </div>
        </div>
        <div
          class="grid grid-cols-3 mt-5 w-full justify-between"
        >
          <span>
            hp
          </span>
          <span
            class="w-full"
          >
            <div
              aria-valuemax="100"
              aria-valuemin="0"
              class="relative h-4 w-full overflow-hidden rounded-full bg-secondary"
              data-max="100"
              data-state="indeterminate"
              role="progressbar"
            >
              <div
                class="h-full w-full flex-1 bg-primary transition-all"
                data-max="100"
                data-state="indeterminate"
                style="transform: translateX(-55%);"
              />
            </div>
          </span>
          <span
            class="text-right"
          >
            45
          </span>
        </div>
      </div>
    </div>
  </main>
</div>
`;
